### PR TITLE
[OpenCast] Add extractor

### PR DESCRIPTION
**Pull request by bwildenhain**
Original found at https://github.com/ytdl-org/youtube-dl/pull/26934
Author: C0D3D3V

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (https://github.com/ytdl-org/youtube-dl/pull/26934)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

[Opencast](https://opencast.org/) is a open source software used  by universities to publish recorded lectures. This allows to download lectures presented by Opencasts' "paella"-player.

### CODE REVIEW

File: yt_dlp/extractor/opencast.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                            mcmedia\.missioncollege\.org|
+                            clases\.odon\.edu\.uy
+                        )'''
+    _UUID_RE = r'[\da-fA-F]{8}-[\da-fA-F]{4}-[\da-fA-F]{4}-[\da-fA-F]{4}-[\da-fA-F]{12}'
+
+    def _call_api(self, host, video_id, path, note=None, errnote=None, fatal=True):
+        return self._download_json(self._API_BASE % (host, video_id), video_id, note=note, errnote=errnote, fatal=fatal)
+
+    def _parse_mediapackage(self, video):
+        tracks = video.get('media', {}).get('track', [])
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:46]: don't use get chains, use try_get instead

File: yt_dlp/extractor/opencast.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            ext = determine_ext(href, None)
+            track_obj = {'url': href}
+
+            transport = track.get('transport')
+
+            if transport == 'DASH' or ext == 'mpd':
+                formats.extend(self._extract_mpd_formats(href, video_id, mpd_id='dash', fatal=False))
+            elif transport == 'HLS' or ext == 'm3u8':
+                formats.extend(
+                    self._extract_m3u8_formats(href, video_id, m3u8_id='hls', entry_protocol='m3u8_native', fatal=False)
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:62]: use _extract_*_formats_and_subtitles just in case

File: yt_dlp/extractor/opencast.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                        track_obj.update({'vcodec': video_encoder.get('type')})
+
+                formats.append(track_obj)
+
+        self._sort_formats(formats)
+
+        result_obj = {'formats': formats}
+
+        if video_id is not None:
+            result_obj.update({'id': video_id})
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:120]: id is supposed to be fatal
may as well stop the extraction if it doesn't exist

File: yt_dlp/extractor/opencast.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                            media\.uct\.ac\.za|
+                            vid\.igb\.illinois\.edu|
+                            cursosabertos\.c3sl\.ufpr\.br|
+                            mcmedia\.missioncollege\.org|
+                            clases\.odon\.edu\.uy
+                        )'''
+    _UUID_RE = r'[\da-fA-F]{8}-[\da-fA-F]{4}-[\da-fA-F]{4}-[\da-fA-F]{4}-[\da-fA-F]{12}'
+
+    def _call_api(self, host, video_id, path, note=None, errnote=None, fatal=True):
+        return self._download_json(self._API_BASE % (host, video_id), video_id, note=note, errnote=errnote, fatal=fatal)
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:43]: path is not used anywhere...just remove it from the signature.

File: yt_dlp/extractor/opencast.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+        search_results = api_json.get('search-results', {})
+        if 'result' not in search_results:
+            raise ExtractorError('Video was not found')
+
+        result_dict = search_results.get('result', {})
+        if not isinstance(result_dict, dict):
+            raise ExtractorError('More than one video was unexpectedly returned.')
+
+        video = result_dict.get('mediapackage', {})
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:189]: why not combine all these and do
```suggestion
        video = self._call_api(host, video_id, '', note='Downloading video JSON')['search-results']['result']['mediapackage']
```

File: yt_dlp/extractor/opencast.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+        result_obj = self._parse_mediapackage(video)
+        return result_obj
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:192]:
```suggestion
        return self._parse_mediapackage(video)
```

File: yt_dlp/extractor/opencast.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+                    https?://(?P<host>%s)/paella/ui/watch.html\?.*?
+                    id=(?P<id>%s)
+                    ''' % (
+        OpencastBaseIE._INSTANCES_RE,
+        OpencastBaseIE._UUID_RE,
+    )
+
+    _API_BASE = 'https://%s/search/episode.json?id=%s'
+
+    _TEST = {
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:161]: make this `_TESTS = [...]`

File: yt_dlp/extractor/opencast.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+                    https?://(?P<host>%s)/engage/ui/index.html\?.*?
+                    epFrom=(?P<id>%s)
+                    ''' % (
+        OpencastBaseIE._INSTANCES_RE,
+        OpencastBaseIE._UUID_RE,
+    )
+
+    _API_BASE = 'https://%s/search/episode.json?sid=%s'
+
+    _TESTS = [
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:206]: formatting for this is sus. Idk why isn't the linter catching this 🤔

File: yt_dlp/extractor/opencast.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            'info_dict': {
+                'id': 'b1a54262-3684-403f-9731-8e77c3766f9a',
+                'title': 'inSTUDIES-Social movements and prefigurative politics in a global perspective',
+            },
+            'playlist_mincount': 8,
+        },
+    ]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:226]: use self._match_valid_url()

File: yt_dlp/extractor/opencast.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+        host = mobj.group('host')
+        video_id = mobj.group('id')
+
+        api_json = self._call_api(host, video_id, '', note='Downloading video JSON')
+
+        search_results = api_json.get('search-results', {})
+        if 'result' not in search_results:
+            raise ExtractorError('Playlist was not found')
+
+        result_list = search_results.get('result', {})
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:236]: again combining all these to 
```suggestion
        result_list = self._call_api(host, video_id, '', note='Downloading video JSON')['search-results']['result']
```

File: yt_dlp/extractor/opencast.py [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+        if isinstance(result_list, dict):
+            result_list = [result_list]
+
+        entries = []
+        for episode in result_list:
+            video = episode.get('mediapackage', {})
+            entries.append(self._parse_mediapackage(video))
+
+        if len(entries) == 0:
+            raise ExtractorError('Playlist has no entries')
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:246]: Do something like this (the formatting is off in the suggestion so make sure to run flake8 on this)
```suggestion
        entries = [
        self._parse_mediapackage(episode['mediapackage'])
        for episode in result_list 
        if episode.get('mediapackage')]
```

File: yt_dlp/extractor/opencast.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+
+        playlist_title = entries[0].get('series')
+
+        result_obj = self.playlist_result(entries, playlist_id=video_id, playlist_title=playlist_title)
+        return result_obj
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:251]:
```suggestion
        return self.playlist_result(entries, playlist_id=video_id, playlist_title=try_get(entries, lambda x:x[0]['series']))
```
Can further be "improved" by doing traverse_obj and going to next entry until series is not found but i personally don't like importing stuff just for such things

File: yt_dlp/extractor/opencast.py (hunk 8 above)
Comment (bwildenhain) [reply yt_dlp/extractor/opencast.py:196]: Thanks a lot for your comments! Would you expected the indentation to be like the changed source in l. 186 to 192?

File: yt_dlp/extractor/opencast.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        if season_id is not None:
+            result_obj.update({'season_id': season_id})
+
+        creator = video.get('creators', {}).get('creator')
+        if creator is not None:
+            result_obj.update({'creator': creator})
+
+        timestamp = parse_iso8601(video.get('start'))
+        if timestamp is not None:
+            result_obj.update({'timestamp': timestamp})
```
Comment (nao20010128nao) [inline yt_dlp/extractor/opencast.py:139]:
```suggestion
        return {
            'id': video_id,
            'formats': formats,
            'title': video.get('title'),
            'series': video.get('seriestitle'),
            'season_id': video.get('series'),
            'creator': traverse_obj(video, ('creators', 'creator')),
            'timestamp': parse_iso8601(video.get('start')),
        }
```

`None` value for mostly any key (exceptions are `id` and either `url` or `formats`) are allowed here

File: yt_dlp/extractor/opencast.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+
+        attachments = video.get('attachments', {}).get('attachment', [])
+        if len(attachments) > 0:
+            thumbnail = attachments[0].get('url')
+            result_obj.update({'thumbnail': thumbnail})
```
Comment (nao20010128nao) [inline yt_dlp/extractor/opencast.py:144]: depending on your intention,

```suggestion
        'thumbnails': [{'url': x} for x in traverse_obj(video, ('attachments', 'attachment', ..., 'url'), default=[])],
```

or

```suggestion
        'thumbnail': traverse_obj(video, ('attachments', 'attachment', ..., 'url'), get_all=False),
```

Also, do not apply as-is. Merge with the above.

**Review by nao20010128nao — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/opencast.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+                formats.extend(self._extract_f4m_formats(href, video_id, f4m_id='hds', fatal=False))
+            elif transport == 'SMOOTH':
+                formats.extend(self._extract_ism_formats(href, video_id, ism_id='smooth', fatal=False))
+            elif ext == 'smil':
+                formats.extend(self._extract_smil_formats(href, video_id, fatal=False))
+            else:
+                if transport is not None:
+                    track_obj.update({'format_note': track.get('transport')})
+                    if transport == 'RTMP':
+                        m_obj = re.search(r'^(?:rtmp://[^/]+/(?P<app>[^/]+))/(?P<playpath>.+)$', href)
```
Comment (nao20010128nao) [inline yt_dlp/extractor/opencast.py:74]:
```suggestion
                        m_obj = re.match(r'(?:rtmp://[^/]+/(?P<app>[^/]+))/(?P<playpath>.+)', href)
```

also I'd prefer to add a group for extension extraction (I won't force you unless the maintainer says the same)

File: yt_dlp/extractor/opencast.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+
+                audio_info = track.get('audio')
+                if audio_info is not None:
+                    if 'bitrate' in audio_info:
+                        track_obj.update({'abr': int_or_none(audio_info.get('bitrate'), 1000)})
+                    if 'samplingrate' in audio_info:
+                        track_obj.update({'asr': int_or_none(audio_info.get('samplingrate'))})
+                    audio_encoder = audio_info.get('encoder', {})
+                    if 'type' in audio_encoder:
+                        track_obj.update({'acodec': audio_encoder.get('type')})
```
Comment (nao20010128nao) [inline yt_dlp/extractor/opencast.py:97]:
```suggestion
                    track_obj.update({
                        'abr': int_or_none(audio_info.get('bitrate'), 1000),
                        'asr': int_or_none(audio_info.get('samplingrate')),
                        'acodec': traverse_obj(audio_info, ('encoder', 'type')),
                    })
```

`None` values are allowed here as well
Same applies to video below

**Review by nao20010128nao — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/opencast.py (hunk 8 above)
Comment (pukkandan) [reply yt_dlp/extractor/opencast.py:196]: both styles are valid for this

File: yt_dlp/extractor/opencast.py (hunk 13 above)
Comment (bwildenhain) [reply yt_dlp/extractor/opencast.py:139]: Thanks, I don't know that

File: yt_dlp/extractor/opencast.py [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+
+import re
+
+from .common import InfoExtractor
+from ..utils import (
+    determine_ext,
+    parse_iso8601,
+    int_or_none,
+    traverse_obj,
+    try_get,
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:12]: sort these

File: yt_dlp/extractor/opencast.py [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+                        )'''
+    _UUID_RE = r'[\da-fA-F]{8}-[\da-fA-F]{4}-[\da-fA-F]{4}-[\da-fA-F]{4}-[\da-fA-F]{12}'
+
+    def _call_api(self, host, video_id, note=None, errnote=None, fatal=True):
+        return self._download_json(self._API_BASE % (host, video_id), video_id, note=note, errnote=errnote, fatal=fatal)
+
+    def _parse_mediapackage(self, video):
+        tracks = try_get(video, lambda x: x['media']['track'])
+
+        video_id = video.get('id')
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:48]: should be fatal

File: yt_dlp/extractor/opencast.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+        formats = []
+        for track in tracks:
+            href = track['url']
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:52]: shouldn't be fatal.
use .get and only append if it exists

File: yt_dlp/extractor/opencast.py [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+                                'play_path': m_obj.group('ext') + ':' + m_obj.group('playpath'),
+                                'rtmp_live': True,
+                                'preference': -2,
+                            }
+                        )
+
+                audio_info = track.get('audio')
+                if audio_info is not None:
+                    track_obj.update({
+                        'abr': int_or_none(audio_info.get('bitrate'), 1000),
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:90]: is it really safe to use `1000` as default here?

File: yt_dlp/extractor/opencast.py [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+                        'asr': int_or_none(audio_info.get('samplingrate')),
+                        'acodec': traverse_obj(audio_info, ('encoder', 'type')),
+                    })
+
+                video_info = track.get('video')
+                if video_info is not None:
+                    track_obj.update({
+                        'resolution': video_info.get('resolution'),
+                        'fps': int_or_none(video_info.get('framerate')),
+                        'vbr': int_or_none(video_info.get('bitrate'), 1000),
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:100]: same
Don't use any random int as default

File: yt_dlp/extractor/opencast.py [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+                'timestamp': 1606208400,
+                'upload_date': '20201124',
+            },
+        }
+    ]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        host = mobj.group('host')
+        video_id = mobj.group('id')
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:149]:
```suggestion
        host, video_id = self._match_valid_url(url).groups()
```

File: yt_dlp/extractor/opencast.py [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+                'title': 'inSTUDIES-Social movements and prefigurative politics in a global perspective',
+            },
+            'playlist_mincount': 8,
+        },
+    ]
+
+    def _real_extract(self, url):
+        mobj = self._match_valid_url(url)
+        host = mobj.group('host')
+        video_id = mobj.group('id')
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:189]:
```suggestion
        host, video_id = self._match_valid_url(url).groups()
```

File: yt_dlp/extractor/opencast.py (hunk 20 above)
Comment (pukkandan) [reply yt_dlp/extractor/opencast.py:90]: that's `scale`, not `default`

File: yt_dlp/extractor/opencast.py [hunk 24]
```diff
@@ -87,7 +92,7 @@ def _parse_mediapackage(self, video):
                 audio_info = track.get('audio')
                 if audio_info is not None:
                     track_obj.update({
-                        'abr': int_or_none(audio_info.get('bitrate'), 1000),
+                        'abr': int_or_none(audio_info.get('bitrate')),
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:95]: yo revert this
sorry i didn't realize this is actually scale and not default parameter for int_or_none 😭

File: yt_dlp/extractor/opencast.py [hunk 25]
```diff
@@ -97,7 +102,7 @@ def _parse_mediapackage(self, video):
                     track_obj.update({
                         'resolution': video_info.get('resolution'),
                         'fps': int_or_none(video_info.get('framerate')),
-                        'vbr': int_or_none(video_info.get('bitrate'), 1000),
+                        'vbr': int_or_none(video_info.get('bitrate')),
```
Comment (Ashish0804) [inline yt_dlp/extractor/opencast.py:105]: same here

File: yt_dlp/extractor/opencast.py (hunk 18 above)
Comment (pukkandan) [reply yt_dlp/extractor/opencast.py:48]: @Ashish0804 It is not necessary that the extractor should make id/title extraction fatal. If the extractor returns without an id, the core program will raise an error instead

**Pull request merged**